src/crypto/bls.h

```diff
@@ -0,0 +1,394 @@
+#define BLS_DATA_SECKEY_SIZE (8 * MCLBN_FP_UNIT_SIZE)
+#define BLS_DATA_PUBKEY_SIZE (8 * MCLBN_FP_UNIT_SIZE * 2 * 3)
+#define BLS_DATA_SIG_SIZE (8 * MCLBN_FP_UNIT_SIZE * 3)
+
+#if 1 // change this to 0 if you want to play around with the BLS12-381
+#define BLS_CURVE mclBn_CurveFp254BNb
+#define BLS_CURVE_ID_SIZE 32
+#define BLS_CURVE_SECKEY_SIZE 32
+#define BLS_CURVE_PUBKEY_SIZE 64
+#define BLS_CURVE_SIG_SIZE 32
+#else
+#define BLS_CURVE mclBls12_CurveFp381
+#define BLS_CURVE_ID_SIZE 32
+#define BLS_CURVE_SECKEY_SIZE 32
+#define BLS_CURVE_PUBKEY_SIZE 96
+#define BLS_CURVE_SIG_SIZE 48
+#endif
+#endif
+
+#ifndef BUILD_BITCOIN_INTERNAL
+typedef std::vector<unsigned char, secure_allocator<unsigned char>> BLSPrivateDataBase;
+template<std::size_t Size> class BLSPrivateData : public BLSPrivateDataBase
+{
+public:
+    BLSPrivateData() : BLSPrivateDataBase(Size) {}
+
+    void setZero()
+    {
+        assign(Size, 0);
+    }
+};
+#endif
+
+template<std::size_t Size> class BLSPublicData : public std::array<unsigned char, Size>
+{
+public:
+    BLSPublicData()
+    {
+        setZero();
+    }
+
+    void setZero()
+    {
+        this->fill(0);
+    }
+};
+
+class CBLSSignature;
+class CBLSPublicKey;
+class CBLSMessageHash;
+
+template<typename Data, size_t _SerSize, typename C>
+class CBLSWrapper
+{
+    template<typename T> friend void* access(T&);
+    template<typename T> friend const void* access(const T&);
+
+protected:
+    Data data;
+    bool fValid{false};
+    mutable uint256 cachedHash;
+
+    inline constexpr size_t GetSerSize() const { return SerSize; }
+
+    virtual bool InternalSetBuf(const void* buf, size_t size) = 0;
+    virtual bool InternalGetBuf(void* buf, size_t size) const = 0;
+    virtual bool InternalEquals(const C& r) const = 0;
+
+public:
+    static const size_t SerSize = _SerSize;
+
+    CBLSWrapper()
+    {
+        UpdateHash();
+    }
+
+    CBLSWrapper(CBLSWrapper&& ref)
+    {
+        std::swap(data, ref.data);
+        std::swap(fValid, ref.fValid);
+        std::swap(cachedHash, ref.cachedHash);
+    }
+
+    bool operator==(const C& r) const
+    {
+        return fValid == r.fValid && InternalEquals(r);
+    }
+    bool operator!=(const C& r) const
+    {
+        return !((*this) == r);
+    }
+
+    bool IsValid() const
+    {
+        return fValid;
+    }
+
+    void SetBuf(const void* buf, size_t size)
+    {
+        if (std::all_of((const char*)buf, (const char*)buf + size, [](char c){ return c == 0; })) {
+            Reset();
+        } else {
+            fValid = InternalSetBuf(buf, size);
+            if (!fValid) {
+                Reset();
+            }
+        }
+        UpdateHash();
+    }
+
+    void Reset()
+    {
+        *((C*)this) = C();
+    }
+
+    void GetBuf(void* buf, size_t size) const
+    {
+        if (!fValid) {
+            memset(buf, 0, size);
+        } else {
+            bool ok = InternalGetBuf(buf, size);
+            assert(ok);
```

Comment:
Never put important logic into the `assert(...)`. You never know if the assert will stay in release builds...
It does in our code, because it's forced somewhere...but nothing we should rely on.